### PR TITLE
Get units refactoring

### PR DESCRIPTION
**Pull request by phlax**

### CODE REVIEW

**Comment by ta2-1** [comment]
@phlax can you tell you plan again?
you have copied `get_step_query()` to `get_units()`... and?

File: pootle/apps/pootle_store/views.py [hunk 1]
```diff
@@ -321,16 +321,32 @@ def get_units(request):
...
         raise Http400(_('Arguments missing.'))
     elif len(pootle_path) > 2048:
         raise Http400(_('Path too long.'))
 
-    User = get_user_model()
-    request.profile = User.get(request.user)
-    limit = request.profile.get_unit_rows()
+    if request.user.is_anonymous():
+        # override the anon user
+        request.user = get_user_model().objects.get_nobody_user()
```
Comment (unho) [inline pootle/apps/pootle_store/views.py]: Which is the problem with previous way to retrieve the user?

File: pootle/apps/pootle_store/views.py [hunk 2]
```diff
@@ -362,11 +378,81 @@ def get_units(request):
...
+            User = get_user_model()
+            try:
+                user = User.objects.get(username=username)
+            except User.DoesNotExist:
+                pass
+
+        if unit_filter == "checks":
+            if checks is not None:
+                checks = checks.split(",")
+            elif 'category' is not None:
```
Comment (unho) [inline pootle/apps/pootle_store/views.py]: `string is not None`?

File: pootle/apps/pootle_store/views.py [hunk 3]
```diff
@@ -362,11 +378,81 @@ def get_units(request):
...
+                if sort_by[0] == '-':
+                    max_field = sort_by[1:]
+                    sort_order = '-sort_by_field'
+                else:
+                    max_field = sort_by
+                    sort_order = 'sort_by_field'
+
+                # It's necessary to use `Max()` here because we can't
+                # use `distinct()` and `order_by()` at the same time
+                # (unless PostreSQL is used and `distinct(field_name)`)
```
Comment (unho) [inline pootle/apps/pootle_store/views.py]: Please fix the copied typo: PostreSQL.

File: pootle/apps/pootle_store/views.py [hunk 4]
```diff
@@ -362,11 +378,81 @@ def get_units(request):
         'store__translation_project__project',
         'store__translation_project__language',
     )
-    step_queryset = get_step_query(request, units_qs)
```
Comment (unho) [inline pootle/apps/pootle_store/views.py]: Is there any reason not to remove `get_step_query` now that you copied all of its code here?

File: pootle/apps/pootle_store/views.py (hunk 1 above)
Comment (phlax) [inline pootle/apps/pootle_store/views.py]: 2 things
- it does an additional lookup for every user for almost no reason
- removing request.profile

File: pootle/apps/pootle_store/views.py (hunk 2 above)
Comment (phlax) [inline pootle/apps/pootle_store/views.py]: yep - that is the equivalent of
```
if checks in request.GET
```

File: pootle/apps/pootle_store/views.py (hunk 4 above)
Comment (phlax) [inline pootle/apps/pootle_store/views.py]: yep - the export view

Comment (phlax) [inline pootle/apps/pootle_store/views.py]: actually that reminds me - i need to fix request.profile on export view

**Comment by phlax** [comment]
> you have copied get_step_query() to get_units()... and?

- factor out all request var manipulation
- factor down the code in get_units
- handle request.vars with form
- shift filter/sort logic to pluggable backend

File: pootle/apps/pootle_store/views.py (hunk 1 above)
Comment (unho) [inline pootle/apps/pootle_store/views.py]: > - it does an additional lookup for every user for almost no reason

If this is about the extra query on `get_step_query` it is now gone that you copied its code here. So why drop `User.get`?


> - removing request.profile

I agree with this.

File: pootle/apps/pootle_store/views.py (hunk 4 above)
Comment (unho) [inline pootle/apps/pootle_store/views.py]: Right, forgot about that.

File: pootle/apps/pootle_store/views.py (hunk 2 above)
Comment (unho) [inline pootle/apps/pootle_store/views.py]: I mean, shouldn't it be `category` instead of `'category'`?

Comment (phlax) [inline pootle/apps/pootle_store/views.py]: @unho youre absolutely right - my apologies

File: pootle/apps/pootle_store/views.py (hunk 1 above)
Comment (phlax) [inline pootle/apps/pootle_store/views.py]: ``User.get`` is the pointless extra query

**Comment by phlax** [comment]
...and then make export code use form/backend etc

File: pootle/apps/pootle_store/views.py (hunk 4 above)
Comment (phlax) [inline pootle/apps/pootle_store/views.py]: > i need to fix request.profile on export view

the export view uses pootle.core.views.set_permissions which sets request.profile - so need to fix in terms of this PR

however this optimization is a pretty valid one so i have PRed an update to set_permissions here #4500

File: pootle/apps/pootle_store/views.py [hunk 5]
```diff
@@ -321,16 +321,32 @@ def get_units(request):
...
+
     if pootle_path is None:
         raise Http400(_('Arguments missing.'))
     elif len(pootle_path) > 2048:
         raise Http400(_('Path too long.'))
 
-    User = get_user_model()
-    request.profile = User.get(request.user)
-    limit = request.profile.get_unit_rows()
+    if request.user.is_anonymous():
```
Comment (ta2-1) [inline pootle/apps/pootle_store/views.py]: this is already landed, is it?

Comment (phlax) [inline pootle/apps/pootle_store/views.py]: no this is for get_units

the other was for pootle.core.views

File: pootle/apps/pootle_store/views.py [hunk 6]
```diff
@@ -350,63 +366,105 @@ def get_units(request):
...
-            uids = uid_list[:count]
-
-    if not units and uids:
-        units = step_queryset.filter(id__in=uids)
-
-    units_by_path = groupby(units, lambda x: x.store.pootle_path)
+            if uids[0] not in uid_list:
+                raise Http404
+            uid = uids[0]
+            index = uid_list.index(uid)
```
Comment (unho) [inline pootle/apps/pootle_store/views.py]: I don't particularly like creating unnecessary variables, so what about replacing this two lines with `index = uid_list.index(uids[0])`?

Comment (phlax) [inline pootle/apps/pootle_store/views.py]: for sure

File: pootle/apps/pootle_store/views.py [hunk 7]
```diff
@@ -350,63 +366,105 @@ def get_units(request):
...
+                units_qs = units_qs.filter(
+                    submitted_on__gt=datetime_obj).distinct()
+
+        if month is not None:
+            [start, end] = get_date_interval(month)
+            units_qs = units_qs.filter(
+                submitted_on__gte=start,
+                submitted_on__lte=end).distinct()
+
+        sort_by = ALLOWED_SORTS[sort_on].get(sort_by_param)
```
Comment (unho) [inline pootle/apps/pootle_store/views.py]: Is there any reason to drop the default to `None` in `get`?

Comment (unho) [inline pootle/apps/pootle_store/views.py]: Stupid question again.

File: pootle/apps/pootle_store/views.py [hunk 8]
```diff
@@ -350,63 +366,105 @@ def get_units(request):
     except Resolver404:
         raise Http404('Unrecognised path')
 
-    units_qs = Unit.objects.get_translatable(
-        user=request.profile,
-        **path_kwargs)
-    units_qs = units_qs.order_by("store", "index")
+    related_selects = (
```
Comment (unho) [inline pootle/apps/pootle_store/views.py]: Another unnecessary variable?

Comment (phlax) [inline pootle/apps/pootle_store/views.py]: in this case i want to keep - it improves formatting - and also takes this list out of the function call

**Comment by unho** [comment]
lgtm

**Comment by ta2-1** [comment]
I haven't got any bugs after clicky testing.
lgtm.

**Pull request merged**